core/app/models/spree/payment_create.rb

```diff
@@ -44,20 +51,33 @@ def build_source
       if source_attributes.present? && payment_method.try(:payment_source_class)
         payment.source = payment_method.payment_source_class.new(source_attributes)
         payment.source.payment_method_id = payment_method.id
-        payment.source.user_id = order.user_id if order
+        if order && payment.source.respond_to?(:user=)
+          payment.source.user = order.user
+        end
       end
     end
 
+    def build_from_wallet_payment_source
+      wallet_payment_source_id = source_attributes.fetch(:wallet_payment_source_id)
+      raise(ActiveRecord::RecordNotFound) if order.user.nil?
+      wallet_payment_source = order.user.wallet.find(wallet_payment_source_id)
+      raise(ActiveRecord::RecordNotFound) if wallet_payment_source.nil?
```

Comment:
Could we add a `find!` method to the `Wallet` class instead? Then raise in that class rather than here.